torax/transport_model/tglf_based_transport_model.py

```diff
@@ -0,0 +1,173 @@
+        q_correction_factor: chex.Numeric,
+        transport: DynamicRuntimeParams,
+        geo: geometry.Geometry,
+        core_profiles: state.CoreProfiles,
+    ) -> TGLFInputs:
+        # Shorthand for the appropriate variables
+        Te = core_profiles.temp_el
+        Ti = core_profiles.temp_ion
+        ne = core_profiles.ne
+
+        # Reference velocity and length, used for normalisation
+        vref = (Te.face_value() / (core_profiles.Ai * CONSTANTS.mp)) ** 0.5
+        lref = geo.Rmin[-1]  # Minor radius at LCFS
+
+        # Temperature gradients
+        Ti_over_Te = Ti.face_value() / Te.face_value()
+        Ate = -lref / Te.face_value() * Te.face_grad()
+        Ati = -lref / Ti.face_value() * Ti.face_grad()
+
+        # Density gradient
+        # Note: nref cancels, as 1/(ne*nref) * (ne_grad * nref) = 1/ne * ne_grad
+        Ane = -lref / ne.face_value() * core_profiles.ne.face_grad()
+
+        # Electron-electron collision frequency
+        # Note: In the TGLF docs, XNUE is mislabelled.
+        # It is actually the electron-electron collision frequency
+        # See https://pyrokinetics.readthedocs.io/en/latest/user_guide/collisions.html
+        Lambda_ee = physics._calculate_lambda_ee(Te, ne)
+        normalised_nu_ee = (4 * jnp.pi * ne * CONSTANTS.qe**4 * Lambda_ee) / (
+            CONSTANTS.me**0.5 * (2 * Te) ** 1.5
+        )
+        nu_ee = normalised_nu_ee / (vref / lref)
+
+        # Safety factor
+        # Need to recalculate since in the nonlinear solver psi has intermediate
+        # states in the iterative solve
+        q, _ = physics.calc_q_from_psi(
+            geo=geo,
+            psi=core_profiles.psi,
+            q_correction_factor=q_correction_factor,
+        )
+        # Shear uses rho_face_norm
+        # TODO: check whether this should be midplane R
+        s_hat = physics.calc_s_from_psi(geo, core_profiles.psi)  # = r/q dq/dr
+
+        # Electron beta
+        p_e = ne * (Te * 1e3)  # ne in m^-3, Te in eV
+        # B_unit = q/r dpsi/dr
+        B_unit = (
+            q / geo.rho_face_norm * jnp.gradient(core_profiles.psi, geo.rho_face_norm)
+        )
+        beta_e = 8 * jnp.pi * p_e / B_unit**2
+
+        # Geometry
+        Rmaj = geo.Rmaj
+        Rmin = geo.Rmin
+        dRmaj = jnp.gradient(geo.Rmaj, geo.rho_face_norm)
+        kappa = geo.elongation_face
+        # Elongation
+        kappa_shear = geo.rho_face_norm / kappa * jnp.gradient(kappa, geo.rho_face_norm)
+        # Triangularity
+        delta = geo.delta_face
+        delta_shear = geo.delta_face * jnp.gradient(geo.delta_face, geo.rho_face_norm)
+
+        # Gyrobohm diffusivity
+        # Used to unnormalise the outputs
+        # TODO: check this definition with Lorenzo/TGLF and ensure correct normalisation
+        chiGB = (
+            (core_profiles.Ai * CONSTANTS.mp) ** 0.5
+            / (CONSTANTS.qe * geo.B0) ** 2
+            * (Ti.face_value() * CONSTANTS.keV2J) ** 1.5
+            / lref
+        )
```

Comment:
Is this definition correct / consistent with TGLF?

Reply:
Use  quasilinear_transport_model.calculate_chiGB , where you can set a b_unit and reference length as input.

For TGLF probably b_unit=Bunit which you need to define (it's not the same as geo.B0), and aminor for the reference length. Should double check

Reply:
Cool, I think this might've been a function added in that I hadn't noticed. I'll recheck the tglf docs for the reference field etc.